File: dev/workshop/js/dim-grid.js
(function () {
  "use strict";

  function initialize(root) {
    if (!root || root.dataset.wdgInitialized === "true") return;
    root.dataset.wdgInitialized = "true";

    var limit = Number(root.dataset.limit);
    var scale = 130 / limit;
    var slopes = {
      actor: Number(root.dataset.actor),
      partner: Number(root.dataset.partner),
      mean: Number(root.dataset.mean),
      within: Number(root.dataset.within)
    };
    var colors = {
      actor: "var(--wdg-actor)",
      partner: "var(--wdg-partner)",
      mean: "var(--wdg-mean)",
      within: "var(--wdg-within)"
    };
    var inputs = {};
    var outputs = {};
    ["actor", "partner", "mean", "within"].forEach(function (name) {
      inputs[name] = root.querySelector('[data-wdg-input="' + name + '"]');
      outputs[name] = root.querySelector('[data-wdg-output="' + name + '"]');
    });

    var plot = root.querySelector("[data-wdg-plot]");
    var point = root.querySelector("[data-wdg-point]");
    var halo = root.querySelector("[data-wdg-halo]");
    var description = root.querySelector("[data-wdg-description]");
    var equationAPIM = root.querySelector('[data-wdg-equation="apim"]');
    var equationDIM = root.querySelector('[data-wdg-equation="dim"]');
    var slopeSummary = root.querySelector("[data-wdg-slopes]");

    function clamp(value) {
      return Math.min(limit, Math.max(-limit, Number(value)));
    }

    function clean(value) {
      return Math.abs(value) < 1e-10 ? 0 : value;
    }

    function format(value, digits) {
      return clean(value).toFixed(digits).replace("-", "−");
    }

    function coefficient(value, following) {
      value = clean(value);
      var magnitude = Math.abs(value).toFixed(2);
      if (!following) return value < 0 ? "−" + magnitude : magnitude;
      return value < 0 ? " − " + magnitude : " + " + magnitude;
    }

    function xPixel(value) { return 160 + value * scale; }
    function yPixel(value) { return 160 - value * scale; }

    function render() {
      var actor = Number(inputs.actor.value);
      var partner = Number(inputs.partner.value);
      var mean = Number(inputs.mean.value);
      var within = Number(inputs.within.value);
      var apimResult = slopes.actor * actor + slopes.partner * partner;
      var dimResult = slopes.mean * mean + slopes.within * within;

      outputs.actor.value = format(actor, 2);
      outputs.partner.value = format(partner, 2);
      outputs.mean.value = format(mean, 2);
      outputs.within.value = format(within, 2);
      equationAPIM.innerHTML =
        coefficient(slopes.actor, false) + " × " + format(actor, 2) +
        coefficient(slopes.partner, true) + " × " + format(partner, 2) +
        " = <strong>" + format(apimResult, 2) + "</strong>";
      equationDIM.innerHTML =
        coefficient(slopes.mean, false) + " × " + format(mean, 2) +
        coefficient(slopes.within, true) + " × " + format(within, 2) +
        " = <strong>" + format(dimResult, 2) + "</strong>";
      slopeSummary.innerHTML =
        "Fitted slopes: <i>b</i><sub>mean</sub> = <i>a</i> + <i>p</i> = " +
        coefficient(slopes.mean, false) +
        "; <i>b</i><sub>dev</sub> = <i>a</i> − <i>p</i> = " +
        coefficient(slopes.within, false);

      point.setAttribute("cx", xPixel(actor));
      point.setAttribute("cy", yPixel(partner));
      halo.setAttribute("cx", xPixel(actor));
      halo.setAttribute("cy", yPixel(partner));
      description.textContent =
        "Actor " + format(actor, 2) + ", partner " + format(partner, 2) +
        ", dyad mean " + format(mean, 2) + ", and member deviation " +
        format(within, 2) + ".";
    }

    function setFromAPIM(actor, partner) {
      actor = clamp(actor);
      partner = clamp(partner);
      inputs.actor.value = actor;
      inputs.partner.value = partner;
      inputs.mean.value = (actor + partner) / 2;
      inputs.within.value = (actor - partner) / 2;
      render();
    }

    function setFromDIM(mean, within) {
      mean = Number(mean);
      within = Number(within);
      var bound = limit - Math.abs(within);
      mean = Math.min(bound, Math.max(-bound, mean));
      bound = limit - Math.abs(mean);
      within = Math.min(bound, Math.max(-bound, within));
      inputs.mean.value = mean;
      inputs.within.value = within;
      inputs.actor.value = mean + within;
      inputs.partner.value = mean - within;
      render();
    }

    inputs.actor.addEventListener("input", function () {
      setFromAPIM(inputs.actor.value, inputs.partner.value);
    });
    inputs.partner.addEventListener("input", function () {
      setFromAPIM(inputs.actor.value, inputs.partner.value);
    });
    inputs.mean.addEventListener("input", function () {
      setFromDIM(inputs.mean.value, inputs.within.value);
    });
    inputs.within.addEventListener("input", function () {
      setFromDIM(inputs.mean.value, inputs.within.value);
    });
    root.querySelector("[data-wdg-reset]").addEventListener("click", function () {
      setFromAPIM(0, 0);
    });
    root.querySelector('[data-wdg-demo="mean"]').addEventListener("click", function () {
      setFromAPIM(1, 1);
    });
    root.querySelector('[data-wdg-demo="within"]').addEventListener("click", function () {
      setFromAPIM(1, -1);
    });

    function setFromPointer(event) {
      var bounds = plot.getBoundingClientRect();
      var svgX = (event.clientX - bounds.left) / bounds.width * 320;
      var svgY = (event.clientY - bounds.top) / bounds.height * 320;
      var actor = Math.round(((svgX - 160) / scale) * 20) / 20;
      var partner = Math.round(((160 - svgY) / scale) * 20) / 20;
      setFromAPIM(actor, partner);
    }

    var dragging = false;
    plot.addEventListener("pointerdown", function (event) {
      dragging = true;
      plot.setPointerCapture(event.pointerId);
      setFromPointer(event);
    });
    plot.addEventListener("pointermove", function (event) {
      if (dragging) setFromPointer(event);
    });
    plot.addEventListener("pointerup", function (event) {
      dragging = false;
      if (plot.hasPointerCapture(event.pointerId)) {
        plot.releasePointerCapture(event.pointerId);
      }
    });
    plot.addEventListener("pointercancel", function () { dragging = false; });

    function drawGrid() {
      var namespace = "http://www.w3.org/2000/svg";
      var lines = root.querySelector("[data-wdg-grid-lines]");
      var labels = root.querySelector("[data-wdg-axis-labels]");

      function addLine(x1, y1, x2, y2, color, opacity, dash) {
        var line = document.createElementNS(namespace, "line");
        line.setAttribute("x1", x1);
        line.setAttribute("y1", y1);
        line.setAttribute("x2", x2);
        line.setAttribute("y2", y2);
        line.setAttribute("stroke", color);
        line.setAttribute("stroke-opacity", opacity);
        line.setAttribute("stroke-width", dash ? "1" : "1.5");
        if (dash) line.setAttribute("stroke-dasharray", dash);
        lines.appendChild(line);
      }

      [-limit, -limit / 2, 0, limit / 2, limit].forEach(function (tick) {
        var dash = tick === 0 ? null : "3 3";
        var opacity = tick === 0 ? "0.62" : "0.22";
        addLine(xPixel(tick), 30, xPixel(tick), 290, colors.actor, opacity, dash);
        addLine(30, yPixel(tick), 290, yPixel(tick), colors.partner, opacity, dash);
        addLine(xPixel(-limit), yPixel(2 * tick + limit), xPixel(limit),
          yPixel(2 * tick - limit), colors.mean, opacity, dash);
        addLine(xPixel(-limit), yPixel(-limit - 2 * tick), xPixel(limit),
          yPixel(limit - 2 * tick), colors.within, opacity, dash);
      });

      function addLabel(text, x, y, color, anchor) {
        var label = document.createElementNS(namespace, "text");
        label.setAttribute("x", x);
        label.setAttribute("y", y);
        label.setAttribute("fill", color);
        label.setAttribute("font-size", "11");
        label.setAttribute("font-weight", "700");
        label.setAttribute("text-anchor", anchor || "middle");
        label.textContent = text;
        labels.appendChild(label);
      }

      addLabel("actor", 286, 151, colors.actor, "end");
      addLabel("partner", 160, 22, colors.partner);
      addLabel("dyad mean", 247, 66, colors.mean);
      addLabel("deviation", 248, 261, colors.within);
    }

    drawGrid();
    setFromAPIM(0, 0);
  }

  function initializeAll() {
    document.querySelectorAll(".workshop-dim-grid").forEach(initialize);
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", initializeAll);
  } else {
    initializeAll();
  }
  document.addEventListener("slidechanged", initializeAll);
})();

File: dev/workshop/R/dim-grid.R
dim_reparameterization_grid <- function(
    model, id = "workshop-dim-grid", predictor_label = "Predictor", limit = 4) {
  if (!inherits(model, "glmmTMB")) {
    stop("`model` must be a fitted glmmTMB exchangeable APIM.", call. = FALSE)
  }
  if (!is.numeric(limit) || length(limit) != 1L || !is.finite(limit) || limit <= 0) {
    stop("`limit` must be one positive number.", call. = FALSE)
  }

  fitted <- .extract_apim_diagram_values(model, type = "exchangeable")
  actor <- unname(fitted$estimates[["actor"]])
  partner <- unname(fitted$estimates[["partner"]])
  slopes <- c(
    actor = actor,
    partner = partner,
    mean = actor + partner,
    within = actor - partner
  )
  if (any(!is.finite(slopes))) {
    stop("The fitted APIM slopes must be finite.", call. = FALSE)
  }

  number <- function(x) {
    trimws(formatC(x, digits = 17, format = "g", decimal.mark = "."))
  }
  escape <- htmltools::htmlEscape
  id <- gsub("[^A-Za-z0-9_-]", "-", id)

  htmltools::HTML(sprintf(
'<div class="workshop-dim-grid" id="%s"
     data-limit="%s" data-actor="%s" data-partner="%s"
     data-mean="%s" data-within="%s">
  <div class="wdg-controls">
    <div class="wdg-toolbar">
      <strong>%s coordinates</strong>
      <button type="button" data-wdg-reset>Reset</button>
    </div>

    <div class="wdg-control-group">
      <div class="wdg-group-title wdg-apim-title">APIM coordinates</div>
      <label class="wdg-control wdg-actor">
        <span><span>Actor, <i>x</i><sub>actor</sub></span><output data-wdg-output="actor">0.0</output></span>
        <input data-wdg-input="actor" type="range" min="-%s" max="%s" step="0.1" value="0">
      </label>
      <label class="wdg-control wdg-partner">
        <span><span>Partner, <i>x</i><sub>partner</sub></span><output data-wdg-output="partner">0.0</output></span>
        <input data-wdg-input="partner" type="range" min="-%s" max="%s" step="0.1" value="0">
      </label>
    </div>

    <div class="wdg-control-group">
      <div class="wdg-group-title wdg-dim-title">DIM coordinates</div>
      <label class="wdg-control wdg-mean">
        <span><span>Dyad mean, <i>x</i><sub>mean</sub></span><output data-wdg-output="mean">0.00</output></span>
        <input data-wdg-input="mean" type="range" min="-%s" max="%s" step="0.05" value="0">
      </label>
      <label class="wdg-control wdg-within">
        <span><span>Within-dyad member deviation, <i>x</i><sub>dev</sub></span><output data-wdg-output="within">0.00</output></span>
        <input data-wdg-input="within" type="range" min="-%s" max="%s" step="0.05" value="0">
      </label>
    </div>

    <div class="wdg-relations">
      <span><i>x</i><sub>mean</sub> = (<i>x</i><sub>actor</sub> + <i>x</i><sub>partner</sub>) / 2</span>
      <span><i>x</i><sub>dev</sub> = (<i>x</i><sub>actor</sub> − <i>x</i><sub>partner</sub>) / 2</span>
    </div>
  </div>

  <div class="wdg-visual">
    <div class="wdg-equations">
      <div><b>APIM</b> <span data-wdg-equation="apim"></span></div>
      <div><b>DIM</b> <span data-wdg-equation="dim"></span></div>
      <div class="wdg-slope-summary" data-wdg-slopes></div>
    </div>

    <div class="wdg-stage">
      <svg class="wdg-plot" data-wdg-plot viewBox="0 0 320 320" role="img"
           aria-labelledby="%s-title %s-description">
        <title id="%s-title">APIM and DIM coordinate grid</title>
        <desc id="%s-description" data-wdg-description>The selected point is the grand-mean reference.</desc>
        <defs><clipPath id="%s-clip"><rect x="30" y="30" width="260" height="260" rx="4"></rect></clipPath></defs>
        <rect x="30" y="30" width="260" height="260" rx="4" fill="none" stroke="currentColor" stroke-opacity="0.32"></rect>
        <g data-wdg-grid-lines clip-path="url(#%s-clip)"></g>
        <g data-wdg-axis-labels></g>
        <circle class="wdg-halo" data-wdg-halo cx="160" cy="160" r="13"></circle>
        <circle class="wdg-point" data-wdg-point cx="160" cy="160" r="6"></circle>
      </svg>

      <div class="wdg-demos">
        <button type="button" data-wdg-demo="mean"><b>Shared dyad level</b><span>Both members +1</span></button>
        <div><i>b</i><sub>mean</sub> = <i>a</i> + <i>p</i></div>
        <button type="button" data-wdg-demo="within"><b>Within-dyad member deviation</b><span>Actor +1, partner −1</span></button>
        <div><i>b</i><sub>dev</sub> = <i>a</i> − <i>p</i></div>
        <p>Drag the dot or move either set of sliders. Both equations give the same fitted change in MVPA.</p>
      </div>
    </div>
  </div>
</div>',
    id, number(limit), number(slopes[["actor"]]), number(slopes[["partner"]]),
    number(slopes[["mean"]]), number(slopes[["within"]]), escape(predictor_label),
    number(limit), number(limit), number(limit), number(limit),
    number(limit), number(limit), number(limit), number(limit),
    id, id, id, id, id, id
  ))
}
